### PR TITLE
 CreateTable in HIVE connector

### PR DESCRIPTION
**Pull request by zhenxiao**
Syntax:

CREATE TABLE table_name (col_name data_type, ...)

This feature is disabled by default for the Hive connector.
Enable by setting 'hive.allow-create-table' to 'true'.

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 1]
```diff
@@ -299,7 +303,56 @@ public ColumnMetadata getColumnMetadata(ConnectorTableHandle tableHandle, Connec
...
     public ConnectorTableHandle createTable(ConnectorSession session, ConnectorTableMetadata tableMetadata)
     {
-        throw new UnsupportedOperationException();
+        if (!allowCreateTable) {
+            throw new PrestoException(PERMISSION_DENIED, "Creating tables is disabled in this Hive catalog");
+        }
+
+        checkArgument(!isNullOrEmpty(tableMetadata.getOwner()), "Table owner is null or empty");
+
+        // get the root directory for the database
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: This comment seems to be on the wrong line

File: presto-main/src/main/java/com/facebook/presto/execution/CreateTableTask.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            throw new SemanticException(TABLE_ALREADY_EXISTS, statement, "Table '%s' already exists", tableName);
+        }
+
+        if (statement.getTableElement().isEmpty()) {
+            throw new SemanticException(MISSING_COLUMN_DEFINITION, statement, "Table '%s' does not have column definition", tableName);
+        }
+
+        List<ColumnMetadata> createTableColumnList = new ArrayList<ColumnMetadata>();
+        int ordinalPosition = 0;
+        Iterator<TableElement> itr = statement.getTableElement().iterator();
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/execution/CreateTableTask.java]: This can use a for-each loop

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 3]
```diff
@@ -335,35 +388,26 @@ public void dropTable(ConnectorTableHandle tableHandle)
         }
     }
 
-    @Override
-    public HiveOutputTableHandle beginCreateTable(ConnectorSession session, ConnectorTableMetadata tableMetadata)
+    private void getColumnInfo(ConnectorTableMetadata tableMetadata,
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: This refactoring makes the diff harder to read. Can you split it into a separate commit?

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java [hunk 4]
```diff
@@ -105,6 +105,17 @@ public void testCreateSampledTableAsSelectLimit()
     }
 
     @Test
+    public void testCreateTable()
```
Comment (electrum) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java]: Does this need to be disabled for connectors that don't implement this? See `TestHiveDistributedQueries` or `TestCassandraDistributed` for an example

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/CreateTableAsSelect.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.tree;
+
+import java.util.Objects;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class CreateTableAsSelect
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/CreateTableAsSelect.java]: Please pull the rename into a separate commit before this one

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/SemanticErrorCode.java [hunk 6]
```diff
@@ -62,5 +62,7 @@
 
     SAMPLE_PERCENTAGE_OUT_OF_RANGE,
 
-    INVALID_SESSION_PROPERTY
+    INVALID_SESSION_PROPERTY,
+
+    MISSING_COLUMN_DEFINITION
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/SemanticErrorCode.java]: Add a trailing comma on the last item

File: presto-main/src/main/java/com/facebook/presto/execution/CreateTableTask.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    @Override
+    public void execute(CreateTable statement, Session session, Metadata metadata, QueryStateMachine stateMachine)
+    {
+        QualifiedTableName tableName = createQualifiedTableName(session, statement.getTable());
+        Optional<TableHandle> tableHandle = metadata.getTableHandle(session, tableName);
+        if (tableHandle.isPresent()) {
+            throw new SemanticException(TABLE_ALREADY_EXISTS, statement, "Table '%s' already exists", tableName);
+        }
+
+        if (statement.getTableElement().isEmpty()) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/execution/CreateTableTask.java]: It should not be possible for this to happen since the grammar requires at least one item, thus we don't need a user error code (it would be a bug in Presto):
```java
checkArgument(!statement.getTableElement().isEmpty(), "no columns for table");

**Comment by zhenxiao** [comment]
@electrum thank you. I've split the PR into 3 commits. comments addressed

File: presto-main/src/main/java/com/facebook/presto/execution/CreateTableTask.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+            throw new SemanticException(MISSING_COLUMN_DEFINITION, statement, "Table '%s' does not have column definition", tableName);
+        }
+
+        List<ColumnMetadata> createTableColumnList = new ArrayList<ColumnMetadata>();
+        int ordinalPosition = 0;
+        Iterator<TableElement> itr = statement.getTableElement().iterator();
+        while (itr.hasNext()) {
+            TableElement tblElement = itr.next();
+            Type columnType = metadata.getType(TypeSignature.parseTypeSignature(tblElement.getType().toLowerCase()));
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/execution/CreateTableTask.java]: The `toLowerCase()` here is not necessary since `TypeSignature` handles this.

This needs handling for an unknown type and the `UNKNOWN` type. Check out `visitCast` in `ExpressionAnalyzer`.

**Comment by zhenxiao** [comment]
@electrum thank you, updated PR

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 9]
```diff
@@ -299,7 +303,55 @@ public ColumnMetadata getColumnMetadata(ConnectorTableHandle tableHandle, Connec
...
+        ImmutableList.Builder<FieldSchema> partitionKeysBuilder = ImmutableList.builder();
+        ImmutableList.Builder<FieldSchema> columnsBuilder = ImmutableList.builder();
+
+        buildColumns(columnNames, columnTypes, columnsBuilder, partitionKeysBuilder, tableMetadata);
+
+        Path targetPath = getTargetPath(schemaName, tableName, schemaTableName);
+
+        SerDeInfo serdeInfo = new SerDeInfo();
+        serdeInfo.setName(tableName);
+        serdeInfo.setSerializationLib(this.hiveStorageFormat.getSerDe());
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: HiveStorageFormat hiveStorageFormat = getHiveStorageFormat(session, this.hiveStorageFormat);

**Comment by zhenxiao** [comment]
@arhimondr thank you, get comment addressed

**Comment by electrum** [comment]
Thanks! I made a few changes and put up a new PR: #2622

I removed the `allow-create-table` setting since we already support creating tables in the Hive connector. Also, creating a table cannot destroy data, so I see no reason for a flag. We only do that for the other features since they can destroy data. If you need this flag, please let me me know and I can put it back.

**Comment by zhenxiao** [comment]
@electrum thank you. Sure, no need for the flag
